### PR TITLE
net: buf: Factor out net_buf_simple to separate source file, add unit tests

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This moves net_buf_simple related code to separate source file.
Having those in separate file makes unit testing easier as simple
network buffers do not use kernel objects, thus can be used
in unit tests without a need for adding any mocks.
Test suite has been added as well to cover some of the API.
The suite is based of the test cases that are already defined for
network buffers.

### CODE REVIEW

**Comment by MariuszSkamra** [comment]
@Thalley @larsgk FYI

**Review by sjanc — DISMISSED** [review]

File: tests/net/mocks/assert.c [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+void assert_post_action(const char *file, unsigned int line)
+{
+	/* ztest_test_pass()/ztest_test_fail() are used to stop the execution
+	 * If this is an unexpected assert (i.e. not following expect_assert())
+	 * calling mock_check_if_assert_expected() will return 'false' as
+	 * a default return value
+	 */
+	if (mock_check_if_assert_expected() == true) {
```
Comment (Thalley) [inline tests/net/mocks/assert.c:28]:
```suggestion
	if (mock_check_if_assert_expected()) {
```

File: tests/net/buf_simple/prj.conf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_ASSERT=y
+CONFIG_ASSERT_LEVEL=2
+CONFIG_ASSERT_VERBOSE=y
+CONFIG_ASSERT_ON_ERRORS=y
+CONFIG_NET_BUF=y
+#CONFIG_NET_BUF_LOG=y
+#CONFIG_NET_BUF_LOG_LEVEL_DBG=y
+#CONFIG_NET_BUF_SIMPLE_LOG=y
```
Comment (Thalley) [inline tests/net/buf_simple/prj.conf:10]: Should these be removed, or uncommented?

File: tests/net/buf_simple/src/main.c [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+{
+	net_buf_simple_add_be64(&buf, u64);
+
+	zassert_mem_equal(be64, net_buf_simple_pull_mem(&buf, sizeof(be64)),
+			  sizeof(be64), "Invalid 64 bits byte order");
+}
+
+ZTEST(net_buf_simple_test_suite, test_net_buf_simple_remove_le16)
+{
+	net_buf_simple_reserve(&buf, 16);
```
Comment (Thalley) [inline tests/net/buf_simple/src/main.c:210]: Why are we reserving 16 octets? 

IMO we should either reserve the number of octets we want to push/remove, or we should use the buffer's size to reserve all (which is what the code is doing now)

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/net/buf_simple/prj.conf (hunk 2 above)
Comment (MariuszSkamra) [reply tests/net/buf_simple/prj.conf:10]: left as it is

File: tests/net/mocks/assert.c (hunk 1 above)
Comment (MariuszSkamra) [reply tests/net/mocks/assert.c:28]: That is the code I copied from `tests/bluetooth/host/host/mock/`, thus i.e. copyrights were left unchanged

File: tests/net/buf_simple/prj.conf (hunk 2 above)
Comment (Thalley) [reply tests/net/buf_simple/prj.conf:10]: Why?

**Review by Thalley — DISMISSED** [review]
Assuming code is just moved without changes, this LGTM

File: tests/net/buf_simple/src/main.c (hunk 3 above)
Comment (MariuszSkamra) [reply tests/net/buf_simple/src/main.c:210]: Yeah, I agree, but it does not make much difference for the test. It was just copied from `tests/net/buf`

**Review by rlubos — DISMISSED** [review]

**Comment by MariuszSkamra** [comment]
Removed `test_framework` tag

**Review by rlubos — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**